remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java

```diff
@@ -166,22 +173,50 @@ public void processMessageReceived(ChannelHandlerContext ctx, RemotingCommand ms
         }
     }
 
-    protected void doBeforeRpcHooks(String addr, RemotingCommand request) {
-        if (rpcHooks.size() > 0) {
-            for (RPCHook rpcHook: rpcHooks) {
-                rpcHook.doBeforeRequest(addr, request);
+    protected Decision preHandle(final HandlerContext context, final RemotingCommand request,
+                                 final CompletableFuture<RemotingCommand> responseFuture) {
+        Decision decision = Decision.CONTINUE;
+        for (Handler handler : handlers) {
+            try {
+                decision = handler.preHandle(context, request, responseFuture);
+            } catch (Throwable e) {
+                RemotingCommand response  = RemotingCommand.createResponseCommand(RemotingSysResponseCode.SYSTEM_ERROR,
+                    e.getMessage());
+                response.setOpaque(request.getOpaque());
+                responseFuture.complete(response);
+                decision = Decision.STOP;
+            }
+
+            if (Decision.STOP == decision) {
+                break;
             }
         }
+        return decision;
     }
 
-    protected void doAfterRpcHooks(String addr, RemotingCommand request, RemotingCommand response) {
-        if (rpcHooks.size() > 0) {
-            for (RPCHook rpcHook: rpcHooks) {
-                rpcHook.doAfterResponse(addr, request, response);
+    protected Decision postHandle(final HandlerContext context, final RemotingCommand request, final RemotingCommand response) {
+        Decision decision = Decision.CONTINUE;
+        for (Handler handler : handlers) {
+            try {
+                decision = handler.postHandle(context, request, response);
+            } catch (Throwable ignore) {
+            }
+            if (Decision.STOP == decision) {
+                break;
             }
         }
+        return decision;
     }
 
+    public void registerRPCHook(RPCHook rpcHook) {
+        if (null != rpcHook) {
+            handlers.add(new HandlerAdaptor(rpcHook));
+        }
```

Comment:
If the RPCHook is not exposed to the end-user, this Adapter is great.

But currently, the widely used DefaultMQProduer or  DefaultMQConsumer, or  DefaultAdminExt have the access to RPCHook. 
It is better to come up with other way to solve it.

Reply:
The flow-control is very useful, So why not apply the ability to the RPCHook, which could benefit the users without adding a new interface.

Reply:
We have two goals: 1) Ensure original RPCHook implementations would continue to work; 2) Better development experience for further developers who develops plugins for RocketMQ